crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2579 @@
+/**<  "info_hash" - in key_schedule_context */
+static const char OSSL_HPKE_INFOHASH_LABEL[] = "\x69\x6e\x66\x6f\x5f\x68\x61\x73\x68";
+/**<  "base_nonce" - base nonce calc label */
+static const char OSSL_HPKE_NONCE_LABEL[] = "\x62\x61\x73\x65\x5f\x6e\x6f\x6e\x63\x65";
+/**<  "exp" - internal exporter secret generation label */
+static const char OSSL_HPKE_EXP_LABEL[] = "\x65\x78\x70";
+/**<  "sec" - external label for exporting secret */
+static const char OSSL_HPKE_EXP_SEC_LABEL[] = "\x73\x65\x63";
+/**<  "key" - label for use when generating key from shared secret */
+static const char OSSL_HPKE_KEY_LABEL[] = "\x6b\x65\x79";
+/**<  "psk_hash" - for hashing PSK */
+static const char OSSL_HPKE_PSK_HASH_LABEL[] = "\x70\x73\x6b\x5f\x68\x61\x73\x68";
+/**<  "secret" - for generating shared secret */
+static const char OSSL_HPKE_SECRET_LABEL[] = "\x73\x65\x63\x72\x65\x74";
+/**<  "dkp_prk" - DeriveKeyPair label */
+static const char OSSL_HPKE_DPK_LABEL[] = "\x64\x6b\x70\x5f\x70\x72\x6b";
+/**<  "candidate" - used in deterministic key gen */
+static const char OSSL_HPKE_CAND_LABEL[] = "\x63\x61\x6e\x64\x69\x64\x61\x74\x65";
+/**<  "sk" - label used in deterministic key gen */
+static const char OSSL_HPKE_SK_LABEL[] = "\x73\x6b";
+
+/*
+ * PEM header/footer for private keys
+ * PEM_STRING_PKCS8INF is just: "PRIVATE KEY"
+ */
+#define PEM_PRIVATEHEADER "-----BEGIN "PEM_STRING_PKCS8INF"-----\n"
+#define PEM_PRIVATEFOOTER "\n-----END "PEM_STRING_PKCS8INF"-----\n"
+
+/* max string len we'll try map to a suite */
+#define OSSL_HPKE_MAX_SUITESTR 38
+
+/* "strength" input to RAND_bytes_ex */
+#define OSSL_HPKE_RSTRENGTH 10
+
+/* an error macro just to make things easier */
+
+/*
+ * @brief info about an AEAD
+ */
+typedef struct {
+    uint16_t            aead_id; /**< code point for aead alg */
+    const char *        name;   /* alg name */
+    size_t              taglen; /**< aead tag len */
+    size_t              Nk; /**< size of a key for this aead */
+    size_t              Nn; /**< length of a nonce for this aead */
+} hpke_aead_info_t;
+
+/*
+ * @brief table of AEADs
+ */
+static hpke_aead_info_t hpke_aead_tab[] = {
+    { 0, NULL, 0, 0, 0 }, /* treat 0 as error so nothing here */
+    { OSSL_HPKE_AEAD_ID_AES_GCM_128, LN_aes_128_gcm, 16, 16, 12 },
+    { OSSL_HPKE_AEAD_ID_AES_GCM_256, LN_aes_256_gcm, 16, 32, 12 },
+#ifndef OPENSSL_NO_CHACHA20
+# ifndef OPENSSL_NO_POLY1305
+    { OSSL_HPKE_AEAD_ID_CHACHA_POLY1305, LN_chacha20_poly1305, 16, 32, 12 },
+# endif
+    { OSSL_HPKE_AEAD_ID_EXPORTONLY, LN_aes_128_gcm, 16, 16, 12 }
+#endif
+};
+
+/*
+ * @brief info about a KEM
+ */
+typedef struct {
+    uint16_t       kem_id; /**< code point for key encipherment method */
+    const char     *keytype; /**< string form of algtype "EC"/"X25519"/"X448" */
+    const char     *groupname; /**< string form of EC group for NIST curves  */
+    int            groupid; /**< NID of KEM */
```

Comment:
Use the group name rather than the nid.. (groupid may not be needed)

Reply:
Come to think of it - this table should not even be needed...

Reply:
I think that table remains require, e.g. for GREASE, and when mapping from codepoint+buffer to EVP_PKEY. The NIDs seemed to be needed when mapping from private buffer to EVP_PKEY when I needed to re-calc the public value, but that could be overtaken by events since - will check.

Reply:
Just thinking out loud - Maybe this stuff should just be a getter that calls the dhkem code instead of a duped table here..

Reply:
did that